workflow.py

```diff
@@ -214,54 +215,62 @@ async def eliminate_subtraction(
     :param run_subprocess: runs a subprocess with error handling
-    :param subtraction: the subtraction to align and eliminate reads against
+    :param subtractions: the subtraction to align and eliminate reads against
     :param subtracted_sam_path: path to the SAM file with subtraction-mapped reads removed
     :param work_path: path to the workflow working directory
     """
     to_subtraction_sam_path = work_path / "to_subtraction.sam"
 
-    # Map reads to the subtraction.
-    await run_subprocess(
-        [
-            "bowtie2",
-            "--local",
-            "--no-unal",
-            "--no-hd",
-            "--no-sq",
-            "-N",
-            "0",
-            "-p",
-            str(proc),
-            "-x",
-            shlex.quote(str(subtraction.bowtie2_index_path)),
-            "-U",
-            str(isolate_fastq_path),
-            "-S",
-            str(to_subtraction_sam_path),
-        ]
-    )
-
-    await run_subprocess(
-        [
-            "./eliminate_subtraction",
-            str(isolate_sam_path),
-            str(to_subtraction_sam_path),
-            str(subtracted_sam_path),
-        ]
-    )
-
-    await aiofiles.os.remove(to_subtraction_sam_path)
-
     subtracted_count = 0
 
-    async with aiofiles.open("subtracted_read_ids.txt", "r") as f:
-        async for line in f:
-            if line:
-                subtracted_count += 1
-
-    logger.info(
-        f"Subtracted {subtracted_count} reads that mapped better to a subtraction."
-    )
+    # iterate over all subtractions
+    for subtraction in subtractions:
+        # Map reads to the subtraction.
+        await run_subprocess(
+            [
+                "bowtie2",
+                "--local",
+                "--no-unal",
+                "--no-hd",
+                "--no-sq",
+                "-N",
+                "0",
+                "-p",
+                str(proc),
+                "-x",
+                shlex.quote(str(subtraction.bowtie2_index_path)),
+                "-U",
+                str(isolate_fastq_path),
+                "-S",
+                str(to_subtraction_sam_path),
+            ]
+        )
+
+        await run_subprocess(
+            [
+                "./eliminate_subtraction",
+                str(isolate_sam_path),
+                str(to_subtraction_sam_path),
+                str(subtracted_sam_path),
+            ]
+        )
+
+        await aiofiles.os.remove(to_subtraction_sam_path)
+
+        async with aiofiles.open("subtracted_read_ids.txt", "r") as f:
+            async for line in f:
+                if line:
+                    subtracted_count += 1
+
+        logger.info(
+            "Subtracted %s reads that mapped better to a subtraction.",
+            subtracted_count
+        )
+
+        # don't want to compare old subtractions,
+        # so we set the next iteration's isolate file
+        # to the current subtracted sam file
```

Comment:
I didn't think of this.

The real optimization would be cutting down the FASTQ input on each iteration. After the first iteration is complete, you know that some reads should not be used as FASTQ input for the next iterations because they have better alignments to the first subtraction.